Sources/audio-devices/AudioDevices.swift

```diff
@@ -1,181 +1,302 @@
+    hasInput = inputChannels > 0
+    hasOutput = outputChannels > 0
+  }
+  
+  func getInputVolume() -> Double? {
+    do {
       var deviceVolume: Float32 = 0.0
-
-      do {
-        try CoreAudioData.getAudioData(
-          id: deviceId,
-          selector: kAudioDevicePropertyVolumeScalar,
-          scope: kAudioDevicePropertyScopeOutput,
-          value: &deviceVolume
-        )
-
-        volume = Double(deviceVolume)
-      } catch {
-        volume = nil
-      }
-    } else {
-      volume = nil
+      try CoreAudioData.getAudioData(
+        id: id,
+        selector: kAudioDevicePropertyVolumeScalar,
+        scope: kAudioDevicePropertyScopeInput,
+        value: &deviceVolume
+      )
+      
+      return Double(deviceVolume)
+    } catch {
+      return nil
     }
   }
-
-  mutating func setVolume(_ newVolume: Double) throws {
-    guard volume != nil else {
-      throw AudioDevices.Error.volumeNotSupported
+  
+  func getOutputVolume() -> Double? {
+    do {
+      var deviceVolume: Float32 = 0.0
+      try CoreAudioData.getAudioData(
+        id: id,
+        selector: kAudioHardwareServiceDeviceProperty_VirtualMasterVolume,
+        scope: kAudioDevicePropertyScopeOutput,
+        value: &deviceVolume
+      )
+      
+      return Double(deviceVolume)
+    } catch {
+      return nil
     }
-
+  }
+  
+  func getVolume() -> Double? {
+    // TODO: need to think how to handle when hasInput && hasOutput
+    if hasInput {
+      return getInputVolume()
+    }
+    if hasOutput {
+      return getOutputVolume()
+    }
+    
+    return nil
+  }
+  
+  func getInputMute() -> Bool? {
+    do {
+      var muteValue: UInt32 = 0
+      try CoreAudioData.getAudioData(
+        id: id,
+        selector: kAudioDevicePropertyMute,
+        scope: kAudioDevicePropertyScopeInput,
+        value: &muteValue
+      )
+      
+      return muteValue != 0
+    } catch {
+      return nil
+    }
+  }
+  
+  func getOutputMute() -> Bool? {
+    do {
+      var muteValue: UInt32 = 0
+      try CoreAudioData.getAudioData(
+        id: id,
+        selector: kAudioDevicePropertyMute,
+        scope: kAudioDevicePropertyScopeOutput,
+        value: &muteValue
+      )
+      
+      return muteValue != 0
+    } catch {
+      return nil
+    }
+  }
+  
+  func getMute() -> Bool? {
+    // TODO: need to think how to handle when hasInput && hasOutput
+    if hasInput {
+      return getInputMute()
+    }
+    if hasOutput {
+      return getOutputMute()
+    }
+    
+    return nil
+  }
+  
+  mutating func setInputVolume(_ newVolume: Double) throws {
     guard (0...1).contains(newVolume) else {
       throw AudioDevices.Error.invalidVolumeValue
     }
-
+    
     var value = Float32(newVolume)
     try CoreAudioData.setAudioData(
       id: id,
       selector: kAudioDevicePropertyVolumeScalar,
+      scope: kAudioDevicePropertyScopeInput,
+      value: &value
+    )
+  }
+  
+  mutating func setOutputVolume(_ newVolume: Double) throws {
```

Comment:
```suggestion
  mutating func setVolume(to newVolume: Double, for type: AudioType = [.input, .output]) throws {
```

Which would be called as:

```swift
.setVolume(to: 0.5, for: .output)
```

Same for `toggleInputMute`.

This would be more swifty and reduce the amount of methods.